Detectors/TPC/calibration/SpacePoints/src/TrackInterpolation.cxx

```diff
@@ -0,0 +1,427 @@
+{
+  // apply track quality cuts (assume different settings for track with and without points in TRD or TOF)
+  const auto& trkTPC = mTPCTrackVecInput->at(matchITSTPC.getRefTPC().getIndex());
+  const auto& trkITS = mITSTrackVecInput->at(matchITSTPC.getRefITS().getIndex());
+  if (hasOuterPoint) {
+    // track has a match in TRD or TOF
+    if (trkTPC.getNClusterReferences() < param::MinTPCNCls ||
+        trkITS.getNumberOfClusters() < param::MinITSNCls) {
+      return false;
+    }
+    if (trkTPC.getChi2() / trkTPC.getNClusterReferences() > param::MaxTPCChi2 ||
+        trkITS.getChi2() / trkITS.getNumberOfClusters() > param::MaxITSChi2) {
+      return false;
+    }
+  } else {
+    // ITS-TPC only track
+    if (trkTPC.getNClusterReferences() < param::MinTPCNClsNoOuterPoint ||
+        trkITS.getNumberOfClusters() < param::MinITSNClsNoOuterPoint) {
+      return false;
+    }
+    if (trkTPC.getChi2() / trkTPC.getNClusterReferences() > param::MaxTPCChi2 ||
+        trkITS.getChi2() / trkITS.getNumberOfClusters() > param::MaxITSChi2) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool TrackInterpolation::interpolateTrackITSTOF(const std::pair<o2::dataformats::EvIndex<>, o2::dataformats::MatchInfoTOF>& matchTOF)
+{
+  // get TPC cluster residuals to ITS-TOF only tracks
+  size_t trkIdx = mTrackData.size() - 1;
+  auto propagator = o2::base::Propagator::Instance();
+  const auto& matchITSTPC = mITSTPCTrackVecInput->at(matchTOF.first.getIndex());
+  const auto& clTOF = mTOFClusterVecInput->at(matchTOF.second.getTOFClIndex());
+  //const int clTOFSec = (TMath::ATan2(-clTOF.getY(), -clTOF.getX()) + o2::constants::math::PI) * o2::constants::math::Rad2Deg * 0.05; // taken from TOF cluster class as there is no const getter for the sector
+  const int clTOFSec = clTOF.getCount();
+  const float clTOFAlpha = o2::utils::Sector2Angle(clTOFSec);
+  const auto& trkTPC = mTPCTrackVecInput->at(matchITSTPC.getRefTPC().getIndex());
+  const auto& trkITS = mITSTrackVecInput->at(matchITSTPC.getRefITS().getIndex());
+  auto trkWork = trkITS.getParamOut();
+  *mCache = {};                                     // reset the cache
+  mTrackData[trkIdx].clIdx.setEvent(mClRes.size()); // reference the first cluster residual belonging to this track
+  //printf("=== New Track with pt = %.3f, nClsTPC = %i ===\n", trkWork.getQ2Pt(), trkTPC.getNClusterReferences());
+
+  // store the TPC cluster positions in the cache
+  float clusterTimeBinOffset = matchITSTPC.getTimeMUS().getTimeStamp() / mTPCTimeBinMUS;
+  for (int iCl = trkTPC.getNClusterReferences(); iCl--;) {
+    uint8_t sector, row;
+    uint32_t clusterIndexInRow;
+    trkTPC.getClusterReference(iCl, sector, row, clusterIndexInRow);
+    const auto& clTPC = trkTPC.getCluster(iCl, *mTPCClusterIdxStruct, sector, row);
+    float clTPCX;
+    std::array<float, 2> clTPCYZ;
+    mFastTransform->Transform(sector, row, clTPC.getPad(), clTPC.getTime() - clusterTimeBinOffset, clTPCX, clTPCYZ[0], clTPCYZ[1]);
```

Comment:
I guess this Transform call will also apply eventual calibrations? What we need here is an ideal transformation (using only DCS info to correct the nominal VDrift)

Reply:
That I don't know? Do you know which method I should use instead? TransformInTimeFrame()?

Reply:
Hi @martenole : TransformInTimeFrame is ideal in the sense that it applies only vdrift, but it does not take into account the time of the vertex. It just scales the time linearly to z, which is used for the track seeding.
What you need is basically the same, but taking into account the time of the vertex. Please contact @sgorbuno to implement the required function.

Reply:
Hi @shahor02 , 
which corrections should be used for this nominal transformation?
In the transformation formula there are:  
 - vDrift = elParam.ZbinWidth * gasParam.DriftV; // cm/timebin
 - t0 = elParam.PeakingTime / elParam.ZbinWidth;

 and the following corrections which are currently set to 0 in O2, but will exist later:

 - tpcAlignmentZ
 - time-of-flight, which uses Z of the primary vertex
 - vdrift correction with the global Y coordinate
 - drift length correction TPCTransform->GetTime0CorrTime();

Reply:
Hi @sgorbuno 
In Run2 we are excluding from the transformed clusters (with respect to which the residuals are calculated) all data driven calibration, taking only the vdrift from the laser corrected for the P,T from DCS.  I would do the same also in Run3 (i.e. using only 1st  2 corrections you mentioned), since the final vdrift params (with dZ and t0 corrections) will be extracted from the same residuals during space point calibration extraction.
Also, I would eliminate the vdrift vs Y correction: its magnitude is much smaller than the space point calibration and can be absorbed to the latter.

Cheers,
 Ruben

Reply:
Hi @sgorbuno 
thanks a lot for implementing this quickly! I have only one question: In comparison to the Transform() method TransformIdeal() does not have the default value of zero for vertexTime.
For the conversion time -> z-position the time of the TPC cluster is used, combined with the time information from the TPC-ITS matching and the mT0 from the TPCFastTransform.
Would it be clearer to put the time information from the TPC-ITS matching as vertexTime when I call the transform function?
See e.g. line 169 here https://github.com/martenole/AliceO2/blob/tpccalib/Detectors/TPC/calibration/SpacePoints/src/TrackInterpolation.cxx
Cheers,
Ole

Reply:
The value of zero is only valid for triggered events. In principle, we could also add the default for TransformIdeal(), but we could also remove it for Transform() to be sure people know what they are doing.
For time frames, you'll have to provide the correct time, and I think the time from the TPC-ITS matching is most exact

Reply:
Hi @martenole 
Oh, i forgot to put the default value to the declaration. I'll add it. 
I think, you are right - you code will be clearer when you give the TPC-ITS matching time as the vertex time.

Reply:
@davidrohr 
perhaps you are right - we better remove the default value. Should I?

Reply:
@sgorbuno 
There might be some other changes necessary then. You will need to change at least also the TPC-ITS matching code from Ruben (and probably also other parts, but this one I know), because they rely on the default value.
See e.g. line 1359 in MatchTPCITS.cxx

Reply:
@sgorbuno, @martenole : the https://github.com/AliceO2Group/AliceO2/pull/2284 calls Transform with track time provided separately, you can go on with removing the default.